File: server/python/database_importer.py
#!/usr/bin/env python3
"""
Database Importer for Large CSV Files
Handles 1GB+ CSV files with streaming and chunked processing
"""

import os
import sys
import json
import hashlib
import logging
from datetime import datetime
from typing import List, Dict, Optional, Generator
import pandas as pd
from pathlib import Path

# Load environment variables from .env file
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

import psycopg2
from psycopg2.extras import RealDictCursor

# Configuration
DB_CONFIG = {
    'host': os.getenv('PGHOST', 'ep-wandering-morning-a5u0szvw.us-east-2.aws.neon.tech'),
    'port': os.getenv('PGPORT', 5432),
    'database': os.getenv('PGDATABASE', 'neondb'),
    'user': os.getenv('PGUSER', 'neondb_owner'),
    'password': os.getenv('PGPASSWORD', 'npg_9dlJKInqoT1w')
}

BATCH_SIZE = 1000
HASH_FILE = 'import_hashes.json'
IMPORT_LOG = 'import_summary.json'

# Logging setup
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

def sanitize_for_sql(value: str) -> str:
    """Escape single quotes for SQL"""
    if value is None:
        return ''
    return str(value).replace("'", "''")

def calculate_row_hash(row: pd.Series) -> str:
    """Calculate hash for a row to detect changes"""
    row_string = '|'.join(str(val) for val in row)
    return hashlib.md5(row_string.encode()).hexdigest()

def load_hashes() -> Dict[str, str]:
    """Load existing hashes for change detection"""
    if Path(HASH_FILE).exists():
        with open(HASH_FILE, 'r') as f:
            return json.load(f)
    return {}

def save_hashes(hashes: Dict[str, str]):
    """Save hashes for future change detection"""
    with open(HASH_FILE, 'w') as f:
        json.dump(hashes, f, indent=2)

def connect_database():
    """Create database connection with SSL for Neon"""
    try:
        conn = psycopg2.connect(
            host=DB_CONFIG['host'],
            port=DB_CONFIG['port'],
            database=DB_CONFIG['database'],
            user=DB_CONFIG['user'],
            password=DB_CONFIG['password'],
            sslmode='require'
        )
        return conn
    except Exception as e:
        logger.error(f"Database connection failed: {e}")
        return None

def execute_sql_file(sql_file: str) -> bool:
    """Execute SQL file against the database"""
    if not Path(sql_file).exists():
        logger.error(f"SQL file not found: {sql_file}")
        return False
    
    conn = connect_database()
    if not conn:
        return False
    
    try:
        with conn.cursor() as cursor:
            with open(sql_file, 'r', encoding='utf-8') as f:
                sql_content = f.read()
            
            logger.info(f"Executing SQL file: {sql_file}")
            cursor.execute(sql_content)
            conn.commit()
            
            logger.info("✅ SQL executed successfully!")
            return True
            
    except Exception as e:
        logger.error(f"SQL execution failed: {e}")
        conn.rollback()
        return False
    finally:
        conn.close()

def generate_term_sql(term: str, row: pd.Series, headers: List[str]) -> List[str]:
    """Generate SQL statements for a single term matching actual schema"""
    statements = []
    
    if not term or pd.isna(term):
        return statements
    
    safe_term = sanitize_for_sql(term.strip())
    
    # Create a mapping of headers to content
    content_map = {}
    for idx, header in enumerate(headers[1:], 1):  # Skip term column
        if idx < len(row):
            content = row.iloc[idx]
            if content and pd.notna(content) and str(content).strip():
                content_map[header.lower().strip()] = sanitize_for_sql(str(content))
    
    # Extract content for main fields
    definition = content_map.get('definition', content_map.get('overview', content_map.get('description', safe_term)))
    short_definition = content_map.get('short_definition', content_map.get('summary', ''))
    
    # Build applications JSON from available content
    applications = {}
    for key, value in content_map.items():
        if key not in ['definition', 'short_definition', 'summary', 'overview', 'description']:
            applications[key] = value
    
    if applications:
        # Use dollar-quoted string for JSON to avoid quote escaping issues
        applications_json = f"$json${json.dumps(applications)}$json$"
    else:
        applications_json = 'null'
    
    # Insert into terms table with actual schema
    statements.append(
        f"INSERT INTO terms (name, definition, short_definition, applications, category_id) VALUES "
        f"('{safe_term}', '{definition}', '{short_definition}', {applications_json}, "
        f"(SELECT id FROM categories WHERE name = 'AI/ML Generated' LIMIT 1)) "
        f"ON CONFLICT (name) DO UPDATE SET "
        f"definition = EXCLUDED.definition, "
        f"short_definition = EXCLUDED.short_definition, "
        f"applications = EXCLUDED.applications, "
        f"updated_at = now();"
    )
    
    return statements

def process_csv_chunks(filename: str, chunk_size: int = BATCH_SIZE) -> Generator:
    """Process CSV file in chunks to handle large files"""
    logger.info(f"Processing CSV file: {filename}")
    
    # Count total rows first
    total_rows = sum(1 for _ in open(filename)) - 1  # Subtract header
    logger.info(f"Total rows to process: {total_rows}")
    
    # Process in chunks
    for chunk in pd.read_csv(filename, chunksize=chunk_size, low_memory=False):
        yield chunk

def main():
    if len(sys.argv) < 2:
        print("""
AI/ML Glossary Database Importer (Python Version)

Usage:
  python database_importer.py <csv-file> [options]
  python database_importer.py --execute <sql-file>

Options:
  --force-all     Import all records (skip change detection)
  --dry-run       Generate SQL without saving
  --chunk-size N  Set chunk size (default: 1000)
  --execute       Execute existing SQL file against database

Examples:
  python database_importer.py aiml2.csv
  python database_importer.py aiml2.csv --force-all
  python database_importer.py aiml2.csv --dry-run
  python database_importer.py --execute aiml_import_2025-06-30_14-52-19.sql
        """)
        sys.exit(1)
    
    # Check for execute mode
    if '--execute' in sys.argv:
        execute_idx = sys.argv.index('--execute')
        if execute_idx + 1 >= len(sys.argv):
            logger.error("SQL file required after --execute")
            sys.exit(1)
        sql_file = sys.argv[execute_idx + 1]
        success = execute_sql_file(sql_file)
        sys.exit(0 if success else 1)
    
    csv_file = sys.argv[1]
    force_all = '--force-all' in sys.argv
    dry_run = '--dry-run' in sys.argv
    
    # Custom chunk size
    chunk_size = BATCH_SIZE
    if '--chunk-size' in sys.argv:
        idx = sys.argv.index('--chunk-size')
        if idx + 1 < len(sys.argv):
            chunk_size = int(sys.argv[idx + 1])
    
    if not Path(csv_file).exists():
        logger.error(f"File not found: {csv_file}")
        sys.exit(1)
    
    # Start processing
    timestamp = datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
    sql_file = f"aiml_import_{timestamp}.sql"
    
    logger.info(f"Starting import from {csv_file}")
    logger.info(f"SQL output file: {sql_file}")
    
    # Load existing hashes for change detection
    existing_hashes = {} if force_all else load_hashes()
    new_hashes = {}
    
    # SQL file header
    sql_statements = [
        "-- AI/ML Glossary Database Import",
        f"-- Generated: {datetime.now().isoformat()}",
        f"-- Source: {csv_file}",
        "",
        "BEGIN;",
        "",
        "-- Ensure AI/ML Generated category exists",
        "INSERT INTO categories (id, name, description) VALUES ",
        "(gen_random_uuid(), 'AI/ML Generated', 'Terms imported from AI/ML processing system') ",
        "ON CONFLICT (name) DO NOTHING;",
        ""
    ]
    
    processed_count = 0
    changed_count = 0
    chunk_count = 0
    
    try:
        # Get headers first
        headers_df = pd.read_csv(csv_file, nrows=0)
        headers = headers_df.columns.tolist()
        
        # Process file in chunks
        for chunk in process_csv_chunks(csv_file, chunk_size):
            chunk_count += 1
            logger.info(f"Processing chunk {chunk_count} ({len(chunk)} rows)")
            
            for idx, row in chunk.iterrows():
                term = row.iloc[0] if len(row) > 0 else None
                
                if not term or pd.isna(term):
                    continue
                
                # Calculate row hash for change detection
                row_hash = calculate_row_hash(row)
                term_str = str(term).strip()
                
                # Check if changed
                if not force_all and term_str in existing_hashes and existing_hashes[term_str] == row_hash:
                    new_hashes[term_str] = row_hash
                    continue
                
                # Generate SQL for this term
                term_sql = generate_term_sql(term, row, headers)
                if term_sql:
                    sql_statements.extend(term_sql)
                    sql_statements.append("")  # Empty line for readability
                    processed_count += 1
                    changed_count += 1
                
                new_hashes[term_str] = row_hash
                
                if processed_count % 100 == 0:
                    logger.info(f"Generated SQL for {processed_count} terms...")
        
        # SQL footer
        sql_statements.extend([
            "COMMIT;",
            "",
            f"-- Import complete: {processed_count} terms processed ({changed_count} new/changed)"
        ])
        
        # Save SQL file
        if not dry_run:
            with open(sql_file, 'w', encoding='utf-8') as f:
                f.write('\n'.join(sql_statements))
            
            # Save hashes for next run
            save_hashes(new_hashes)
            
            # Save summary
            summary = {
                "input_file": csv_file,
                "sql_file": sql_file,
                "total_terms": len(new_hashes),
                "processed_terms": processed_count,
                "changed_terms": changed_count,
                "timestamp": datetime.now().isoformat()
            }
            
            with open(IMPORT_LOG, 'w') as f:
                json.dump(summary, f, indent=2)
        
        logger.info(f"✅ Import preparation complete!")
        logger.info(f"   Total terms: {len(new_hashes)}")
        logger.info(f"   Changed/new terms: {changed_count}")
        logger.info(f"   SQL file: {sql_file}")
        
        if not dry_run:
            print(f"\n📋 Next steps:")
            print(f"1. Review the SQL file: {sql_file}")
            print(f"2. Execute against your database:")
            print(f"   psql -d aiglossary -f {sql_file}")
        
    except Exception as e:
        logger.error(f"Error processing file: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/database_importer.js
#!/usr/bin/env node

/**
 * AI/ML Glossary Database Importer
 *
 * Works with the existing processor outputs to import completed data to PostgreSQL
 * Integrates seamlessly with the existing 6-processor system
 */

import { promises as fs } from 'node:fs';
import path from 'node:path';

// ── Configuration ─────────────────────────────────────────────────────────────

const CONFIG = {
  // Database settings
  DB_CONFIG: {
    host: process.env.DB_HOST || 'localhost',
    port: process.env.DB_PORT || 5432,
    database: process.env.DB_NAME || 'aiglossary',
    user: process.env.DB_USER || 'postgres',
    password: process.env.DB_PASSWORD || 'password',
  },

  // Import settings
  BATCH_SIZE: 100,
  HASH_FILE: 'import_hashes.json',
  IMPORT_LOG: 'import_summary.json',
};

// ── Logging ───────────────────────────────────────────────────────────────────

function log(level, message, data = null) {
  const timestamp = new Date().toISOString();
  const logMessage = `${timestamp} - ${level.toUpperCase()} - ${message}`;
  console.log(logMessage);

  if (data) {
    console.log(JSON.stringify(data, null, 2));
  }
}

// ── CSV Parser ────────────────────────────────────────────────────────────────

function parseCSVLine(line) {
  const result = [];
  let current = '';
  let inQuotes = false;

  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    const nextChar = line[i + 1];

    if (char === '"' && !inQuotes) {
      inQuotes = true;
    } else if (char === '"' && inQuotes && nextChar === '"') {
      current += '"';
      i++; // Skip next quote
    } else if (char === '"' && inQuotes) {
      inQuotes = false;
    } else if (char === ',' && !inQuotes) {
      result.push(current);
      current = '';
    } else {
      current += char;
    }
  }

  result.push(current);
  return result;
}

// ── File Loaders ──────────────────────────────────────────────────────────────

async function loadCSV(filename) {
  try {
    log('info', `Loading CSV file: ${filename}`);
    const content = await fs.readFile(filename, 'utf-8');
    const lines = content.split('\n').filter((line) => line.trim());

    if (lines.length < 2) {
      throw new Error('CSV file appears to be empty or has no data rows');
    }

    const headers = parseCSVLine(lines[0]);
    const rows = [];

    // Process in chunks to avoid memory issues
    const chunkSize = 1000;
    for (let i = 1; i < lines.length; i += chunkSize) {
      const chunk = lines.slice(i, i + chunkSize);
      const parsedChunk = chunk.map((line) => parseCSVLine(line));
      rows.push(...parsedChunk);

      if (i % 10000 === 0) {
        log('info', `Processed ${i} rows...`);
      }
    }

    log('info', `CSV loaded: ${rows.length} rows, ${headers.length} columns`);
    return { headers, rows };
  } catch (error) {
    log('error', `Failed to load CSV ${filename}: ${error.message}`);
    throw error;
  }
}

// ── Database SQL Generation ───────────────────────────────────────────────────

function sanitizeForSQL(value) {
  if (value === null || value === undefined) return '';
  return String(value).replace(/'/g, "''");
}

function generateTermSQL(term, headers, record, isArray = false) {
  const statements = [];

  if (!term || !term.trim()) {
    return statements; // Skip empty terms
  }

  const safeTerm = sanitizeForSQL(term.trim());

  // 1. Create basic term entry
  statements.push(
    `INSERT INTO basic_terms (id, term, category_id) VALUES ` +
      `(gen_random_uuid(), '${safeTerm}', ` +
      `(SELECT id FROM categories WHERE name = 'AI/ML Generated' LIMIT 1)) ` +
      `ON CONFLICT (term) DO NOTHING;`
  );

  // 2. Create enhanced term with all sections
  const sections = [];

  if (isArray) {
    // CSV format - use headers with row data
    for (let i = 1; i < headers.length && i < record.length; i++) {
      const sectionName = headers[i];
      const sectionContent = record[i];

      if (sectionContent?.trim()) {
        sections.push(`'${sanitizeForSQL(sectionName)}': '${sanitizeForSQL(sectionContent)}'`);
      }
    }
  } else {
    // JSON format - use object properties
    for (const [key, value] of Object.entries(record)) {
      if (key !== headers[0] && value && String(value).trim()) {
        // Skip term column
        sections.push(`'${sanitizeForSQL(key)}': '${sanitizeForSQL(String(value))}'`);
      }
    }
  }

  if (sections.length > 0) {
    statements.push(
      `INSERT INTO enhanced_terms (id, term_id, content) VALUES ` +
        `(gen_random_uuid(), ` +
        `(SELECT id FROM basic_terms WHERE term = '${safeTerm}' LIMIT 1), ` +
        `'{${sections.join(', ')}}') ` +
        `ON CONFLICT (term_id) DO UPDATE SET content = EXCLUDED.content;`
    );
  }

  return statements;
}

// ── Main Import Functions ─────────────────────────────────────────────────────

async function generateSQL(headers, data, isJSON = false) {
  log('info', 'Generating SQL statements for database import');

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
  const sqlFile = `aiml_import_${timestamp}.sql`;

  const sqlStatements = [
    '-- AI/ML Glossary Database Import',
    `-- Generated: ${new Date().toISOString()}`,
    `-- Format: ${isJSON ? 'JSON' : 'CSV'}`,
    `-- Total records: ${data.length}`,
    '',
    'BEGIN;',
    '',
    '-- Ensure AI/ML Generated category exists',
    `INSERT INTO categories (id, name, description) VALUES `,
    `(gen_random_uuid(), 'AI/ML Generated', 'Terms imported from AI/ML processing system') `,
    `ON CONFLICT (name) DO NOTHING;`,
    '',
  ];

  let processedCount = 0;

  for (let i = 0; i < data.length; i++) {
    const record = data[i];
    const term = isJSON ? record[headers[0]] : record[0];

    if (!term || !term.trim()) continue;

    const termSQL = generateTermSQL(term, headers, record, !isJSON);
    sqlStatements.push(...termSQL);
    sqlStatements.push(''); // Empty line for readability

    processedCount++;

    if (processedCount % 100 === 0) {
      log('info', `Generated SQL for ${processedCount} terms...`);
    }
  }

  sqlStatements.push('COMMIT;');
  sqlStatements.push('');
  sqlStatements.push(`-- Import complete: ${processedCount} terms processed`);

  // Save SQL file
  const sqlContent = sqlStatements.join('\n');
  await fs.writeFile(sqlFile, sqlContent, 'utf-8');

  log('info', `Generated SQL import file: ${sqlFile} (${processedCount} terms)`);
  return { sqlFile, processedCount };
}

// ── Main Import Process ───────────────────────────────────────────────────────

async function importFile(inputFile, _options = {}) {
  try {
    log('info', `Starting import process for: ${inputFile}`);

    const ext = path.extname(inputFile).toLowerCase();
    const isJSON = ext === '.json';

    // Load data
    let headers, data;
    if (isJSON) {
      // JSON support would go here
      throw new Error('JSON support not implemented yet. Please use CSV files.');
    } else {
      const result = await loadCSV(inputFile);
      headers = result.headers;
      data = result.rows;
    }

    if (!headers.length || !data.length) {
      throw new Error('No data found in input file');
    }

    // Generate SQL
    const sqlResult = await generateSQL(headers, data, isJSON);

    // Create summary
    const summary = {
      inputFile,
      inputType: isJSON ? 'JSON' : 'CSV',
      totalRecords: data.length,
      totalFields: headers.length,
      sqlFile: sqlResult.sqlFile,
      processedTerms: sqlResult.processedCount,
      timestamp: new Date().toISOString(),
    };

    // Save import summary
    await fs.writeFile(CONFIG.IMPORT_LOG, JSON.stringify(summary, null, 2));

    log('info', 'Import preparation completed successfully', summary);

    console.log('\n📋 Next Steps:');
    console.log(`1. Review the generated SQL file: ${sqlResult.sqlFile}`);
    console.log(`2. Execute against your database:`);
    console.log(`   psql -d aiglossary -f ${sqlResult.sqlFile}`);
    console.log(`3. Check import summary: ${CONFIG.IMPORT_LOG}`);

    return summary;
  } catch (error) {
    log('error', `Import failed: ${error.message}`);
    throw error;
  }
}

// ── CLI Interface ─────────────────────────────────────────────────────────────

async function main() {
  const args = process.argv.slice(2);

  if (args.includes('--help') || args.includes('-h') || args.length === 0) {
    console.log(`
AI/ML Glossary Database Importer

Usage:
  node database_importer.js <input-file> [options]

Supported Input:
  CSV files from any of the 6 processors

Options:
  --help, -h           Show this help

Environment Variables:
  DB_HOST              Database host (default: localhost)
  DB_PORT              Database port (default: 5432)
  DB_NAME              Database name (default: aiglossary)
  DB_USER              Database user (default: postgres)
  DB_PASSWORD          Database password

Examples:
  # Import from any processor output
  node database_importer.js aiml2.csv

Integration with Existing Processors:
  # Step 1: Run any processor
  python aimlv2_csv.py --mode topdown
  node aimlv2_simple.js --csv --topdown
  
  # Step 2: Import the results
  node database_importer.js aiml2.csv
        `);
    return;
  }

  try {
    // Get input file
    const inputFile = args.find((arg) => !arg.startsWith('--'));
    if (!inputFile) {
      throw new Error('Please specify an input file (CSV)');
    }

    // Start import
    await importFile(inputFile);
  } catch (error) {
    log('error', `Fatal error: ${error.message}`);
    process.exit(1);
  }
}

// ── Entry Point ───────────────────────────────────────────────────────────────

main().catch(console.error);
